File: c++/array/palindrome-number.cpp
/*回文数
判断一个整数是否是回文数。回文数是指正序（从左向右）和倒序（从右向左）读都是一样的整数。

示例 1:
输入: 121
输出: true

示例 2:
输入: -121
输出: false
解释: 从左向右读, 为 -121 。 从右向左读, 为 121- 。因此它不是一个回文数。

示例 3:
输入: 10
输出: false
解释: 从右向左读, 为 01 。因此它不是一个回文数。

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/palindrome-number
*/

/*思路一：逐位比较
*/
class Solution {
public:
    bool isPalindrome(int x) {
        if (x < 0)
            return false;
        int len = 1;
        while(x / len >= 10)
            len *= 10;
        while(x > 0){
            int left = x / len;
            int right = x % 10;
            if (left != right)
                return false;
            x = x % len / 10; //去掉left和right元素
            len /= 100; //去掉了两个元素
        }
        return true;
    }
};

/*思路二：数字全部翻转
翻转数字，与原数字比较
*/
class Solution {
public:
    bool isPalindrome(int x) {
        if (x < 0)
            return false;
        long res = 0;
        long divide = x;
        while(divide > 0){
            res = res * 10 + divide % 10;
            divide /= 10;
        }
        return x == res;
    }
};

/*思路三：数字一半翻转
*/
class Solution {
public:
    bool isPalindrome(int x) {
        if (x < 0)
            return false;
        if (x !=0 && x % 10 == 0) //要考虑末尾有0的情况，如：10，不满足下面的循环，单独拿出来讨论
            return false; 
        int reverse = 0;
        while(x > reverse){
            reverse = reverse * 10 + x % 10;
            x /= 10;
        }
        return reverse == x || reverse / 10 == x; //如输入的x为奇数位，则此时reverse比x多一位（中间位）
    }
};

/*思路四：转换为字符串
转换为字符串，首尾两指针比较字符
*/
class Solution {
public:
    bool isPalindrome(int x) {
        if (x < 0)
            return false;
        string temp;
        temp = to_string(x);
        int i = 0;
        int j = temp.size()-1;
        while(i < j){
            if (temp[i] != temp[j])
                return false;
            i++;
            j--;
        }
        return true;
    }
};

File: python3/array/palindrome-number.py
'''回文数
判断一个整数是否是回文数。回文数是指正序（从左向右）和倒序（从右向左）读都是一样的整数。

示例 1:

输入: 121
输出: true
示例 2:

输入: -121
输出: false
解释: 从左向右读, 为 -121 。 从右向左读, 为 121- 。因此它不是一个回文数。
示例 3:

输入: 10
输出: false
解释: 从右向左读, 为 01 。因此它不是一个回文数。
进阶:

你能不将整数转为字符串来解决这个问题吗？

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/palindrome-number
'''

'''
思路一：通过取整和取余操作获取整数中对应的数字进行比较。
好处：不需要考虑整数溢出的问题。
时间复杂度：O(lgn)。
空间复杂度：O(1)。
'''
class Solution:
    def isPalindrome(self, x):
        if x >= 0 and x <= 9:
            return True
        if x < 0:
            return False
        div = 1
        while(x // div >= 10):
            div *= 10
        while x != 0:
            left = x // div
            right = x % 10
            if left != right:
                return False
            x = x % div // 10
            div //= 100
        return True
    
'''
思路二：将数字进行对折后看能否一一对应，即取出后半段数字进行翻转。
注意：注意数据为10的倍数时。
时间复杂度：O(lgn)。
空间复杂度：O(1)。
'''
class Solution:
    def isPalindrome(self, x):
        if x < 0 or x % 10 == 0 and x // 10 != 0:
            return False
        res = 0
        while (x > res):
            res = res * 10 + x % 10
            x //= 10
        return x == res or x == res // 10

'''
思路三：将整数转为字符串。反转字符串，跟原字符串比较。
'''
class Solution:
    def isPalindrome(self, x):
        if x < 0:
            return False
        tmp = str(x)
        reverse = tmp[::-1]
        return tmp == reverse

'''
思路四：将整数转为字符串。字符串首尾相应位数进行比较。
'''
class Solution:
    def isPalindrome(self, x):
        if x < 0:
            return False
        tmp = str(x)
        i = 0
        j = len(tmp)-1
        while(i < j):
            if tmp[i] != tmp[j]:
                return False
            i += 1
            j -= 1
        return True